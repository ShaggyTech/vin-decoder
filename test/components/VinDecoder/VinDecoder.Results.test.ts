import { mount } from '@vue/test-utils';
/* Component */
import VinDecoderResults from '~/components/VinDecoder/Results.vue';
/* Mock API Data */
import { mockRawResults } from '~/test/__mocks__/mockDecodeVinValuesExtendedResults';

const factory = (options: object) => {
  return mount(VinDecoderResults, {
    ...options,
  });
};

describe('VinDecoderResults Component Tests', () => {
  test('component is hidden if rawResults prop is empty', () => {
    const wrapper = factory({
      sync: false,
    });

    expect(wrapper.vm.$el.textContent).toEqual('');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('if loading prop is true, a skeleton loader is shown in place of the results card', () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        loading: true,
      },
    });

    expect(wrapper.find('.results-loader').exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('given rawResults, component renders the correct title and subtitle', async () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        rawResults: mockRawResults,
      },
    });

    await wrapper.vm.$nextTick();
    const { Make, Model, ModelYear, VIN } = mockRawResults;
    const titleItems = { Make, Model, ModelYear };

    const titleText = wrapper.find('.results-card__title').text();
    Object.keys(titleItems).forEach((key) =>
      expect(titleText).toContain(titleItems[key])
    );

    const subtitleText = wrapper.find('.results-card__subtitle').text();
    expect(subtitleText).toContain(`${VIN}`);

    expect(wrapper.element).toMatchSnapshot();
  });

  test('given rawResults, nullish or non-applicable values are filtered out of the results object ref', async () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        rawResults: mockRawResults,
      },
    });

    await wrapper.vm.$nextTick();

    const { Make, Model, ModelYear, VIN } = mockRawResults;
    const validResults = { Make, Model, ModelYear, VIN };

    const resultsRef = wrapper.vm.$data.results;
    expect(resultsRef).toEqual(validResults);
  });

  test('watcher updates filtered results when rawResults changes', async () => {
    const wrapper = factory({
      sync: false,
    });

    await wrapper.vm.$nextTick();

    const { Make, Model, ModelYear, VIN } = mockRawResults;
    const validResults = { Make, Model, ModelYear, VIN };

    // start with empty rawResults
    let resultsRef = wrapper.vm.$data.results;
    expect(resultsRef).toEqual(null);

    // add rawResults
    await wrapper.setProps({ rawResults: mockRawResults });
    resultsRef = wrapper.vm.$data.results;
    expect(resultsRef).toEqual(validResults);

    // remove rawResults
    await wrapper.setProps({ rawResults: null });
    resultsRef = wrapper.vm.$data.results;
    expect(resultsRef).toEqual(null);
  });
});
