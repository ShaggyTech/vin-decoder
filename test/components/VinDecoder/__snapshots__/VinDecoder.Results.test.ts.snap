// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VinDecoderResults Component Tests component is hidden if rawResults prop is empty 1`] = `
<transition-stub
  name="slide-y-transition"
>
  <!---->
</transition-stub>
`;

exports[`VinDecoderResults Component Tests given rawResults, component renders the correct title and subtitle 1`] = `
<transition-stub
  name="slide-y-transition"
>
  <div
    class="results-card v-card v-card--raised v-sheet theme--light"
    style="max-width: 600px;"
  >
    <div
      class="v-card__title results-card__title mx-auto px-2 title"
    >
      <hr
        aria-orientation="horizontal"
        class="v-divider theme--light"
        role="separator"
      />
       
      <span
        class="mx-auto px-4"
      >
        
        2020 Volkswagen Jetta
        
      
      </span>
       
      <hr
        aria-orientation="horizontal"
        class="v-divider theme--light"
        role="separator"
      />
    </div>
     
    <div
      class="v-card__subtitle results-card__subtitle mx-auto px-2 title text-center"
    >
      
      TESTVIN
      
      <hr
        aria-orientation="horizontal"
        class="v-divider theme--light"
        role="separator"
      />
    </div>
     
    <div
      class="v-card__text"
    >
      <div
        class="v-data-table results-card__table v-data-table--fixed-header theme--light"
        max-width="600"
      >
        <div
          class="v-data-table__wrapper"
          style="height: 300px;"
        >
          <table>
            <thead>
              <tr>
                <th
                  class="text-left"
                >
                  Name
                </th>
                 
                <th
                  class="text-left"
                >
                  Value
                </th>
              </tr>
            </thead>
             
            <tbody>
              <tr>
                <td>
                  Make
                </td>
                 
                <td>
                  Volkswagen
                </td>
              </tr>
              <tr>
                <td>
                  Model
                </td>
                 
                <td>
                  Jetta
                </td>
              </tr>
              <tr>
                <td>
                  ModelYear
                </td>
                 
                <td>
                  2020
                </td>
              </tr>
              <tr>
                <td>
                  VIN
                </td>
                 
                <td>
                  TESTVIN
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</transition-stub>
`;

exports[`VinDecoderResults Component Tests if loading prop is true, a skeleton loader is shown in place of the results card 1`] = `
<transition-stub
  name="slide-y-transition"
>
  <div
    aria-busy="true"
    aria-live="polite"
    class="v-skeleton-loader results-loader v-skeleton-loader--is-loading theme--light"
    role="alert"
    style="max-width: 600px;"
  >
    <div
      class="v-skeleton-loader__card-heading v-skeleton-loader__bone"
    >
      <div
        class="v-skeleton-loader__heading v-skeleton-loader__bone"
      />
    </div>
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
    <div
      class="v-skeleton-loader__text v-skeleton-loader__bone"
    />
  </div>
</transition-stub>
`;
